[BUG] `BaseDetector` - Incorrect docstring and tests for `segments_to_change_points`

satvshr opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
fixes #7490


#### What does this implement/fix? Explain your changes.
Removed `segments_to_change_points` and the associated tests, integrated handling numpy arrays into `_coerce_to_df`.

fkiraly reviewed (changes requested):
Thanks for the contribution!

We cannot merge this as it is not a pure refacor and not downwards compatible:

* `segments_to_change_points` gets removed - why do you want to do that?
* it looks like the default for `predict_points` will now break, because the format returned is segments.

satvshr commented:
> * it looks like the default for `predict_points` will now break, because the format returned is segments.

I understood that after making the PR and I was going to edit it but I had a few questions again, what's the point of having `segments_to_change_points` except numpy input when it is only being used in one place?

fkiraly commented:
> what's the point of having `segments_to_change_points` except numpy input when it is only being used in one place?

It is an abstraction with a clear delination, it converts a container of one type (segments) into one of another type (point events).

Personally, I would have made it private though.

satvshr commented:
Made changes to the `_segments_to_change_points` function, removed changes in `_coerce_to_df`, and converted segments_to_change_points to a private function by adding a backspace to it(I assume that is how its done?)

satvshr commented:
A summary of the problem to avoid confusion:
Apart from the original issue, where it was requested to include the numpy dependency in the function, the function itself does not work when the example code provided in the docstring is executed. However, the test for this function passes for some unknown reason.

Additionally, my PR seems to be failing a test for a function that I did not edit, which is even more confusing. Any help would be greatly appreciated. The 2 errors:
```
FAILED sktime/detection/tests/test_all_detectors.py::TestAllDetectors::test_predict_points[HMM-1] - AttributeError: 'RangeIndex' object has no attribute 'left'

ERROR sktime/detection/tests/test_base.py - Failed: In test_sparse_segments_to_dense: function uses no argument 'segments'
```

satvshr commented:
> The function that is failing calls the function that you edited.

I tried to find the call to that function because that is what my first assumption was, but I just could not. 

 > Tha function that is failing calls the function that you edited.
>
> I would recommend to pause this PR for now, as it feels to me that the original specification was ill-defined, i.e., it is unclear > what the intended target state should be. Not your fault, apologies for not catching this earlier.

That's alright, but I'd like to point out that the function which is currently there does not work, if you use the docstring example given and input it to the function, it would throw an error.

satvshr commented:
Also solves #7712

fkiraly commented:
@satvshr, can you kindly point exactly to the test that you think should be failing?

We just ran all the tests for the 0.36.0 release, and nothing related failed.

satvshr commented:
> @satvshr, can you kindly point exactly to the test that you think should be failing?
> 
> We just ran all the tests for the 0.36.0 release, and nothing related failed.

You see no failures as `@pytest.mark.xfail(reason="Changed interface, need to update test")` skips the test, if that line was not there the test would fail.

fkiraly reviewed (approved):
Ah, I see! The test was already skipped!

Makes sense now.
